homeassistant/components/combined_energy/sensor.py

```diff
@@ -0,0 +1,485 @@
+
+    @abstractmethod
+    def _to_native_value(self, raw_value: Any) -> int | float | None:
+        """Convert non-none raw value into usable sensor value."""
+
+    @property
+    def native_value(self) -> int | float | None:
+        """Return the state of the sensor."""
+        if (raw_value := self._raw_value) is not None:
+            return self._to_native_value(raw_value)
+        return None
+
+
+class GenericSensor(CombinedEnergyReadingsSensor):
+    """Sensor that returns the last value of a sequence of readings."""
+
+    _attr_suggested_display_precision = 2
+
+    def _to_native_value(self, raw_value: Any) -> float:
+        """Convert non-none raw value into usable sensor value."""
+        if isinstance(raw_value, Sequence):
+            raw_value = raw_value[-1]
+        return float(round(raw_value, self.suggested_display_precision))
+
+
+class EnergySensor(CombinedEnergyReadingsSensor):
+    """Sensor for energy readings."""
+
+    _attr_suggested_display_precision = 2
+
+    @property
+    def last_reset(self) -> datetime | None:
+        """Last time the data was reset."""
+        if device_readings := self.device_readings:
+            # mypy is struggling with a Pydantic model here, the cast isn't technically required
+            return cast(datetime | None, device_readings.range_start)
+        return None
+
+    def _to_native_value(self, raw_value: Any) -> float:
+        """Convert non-none raw value into usable sensor value."""
+        value = sum(raw_value)
+        return float(round(value, self.suggested_display_precision))
+
+
+class PowerSensor(CombinedEnergyReadingsSensor):
+    """Sensor for power readings."""
+
+    _attr_suggested_display_precision = 2
+
+    def _to_native_value(self, raw_value: Any) -> float:
+        """Convert non-none raw value into usable sensor value."""
+        return float(round(raw_value, self.suggested_display_precision))
+
+
+class PowerFactorSensor(CombinedEnergyReadingsSensor):
+    """Sensor for power factor readings."""
+
+    _attr_suggested_display_precision = 1
+
+    def _to_native_value(self, raw_value: Any) -> float:
+        """Convert non-none raw value into usable sensor value."""
+        # The API expresses the power factor as a fraction convert to %
+        return float(round(raw_value[-1] * 100, self.suggested_display_precision))
+
+
+class WaterVolumeSensor(CombinedEnergyReadingsSensor):
+    """Sensor for water volume readings."""
+
+    def _to_native_value(self, raw_value: Any) -> int:
+        """Convert non-none raw value into usable sensor value."""
+        return int(round(raw_value[-1], 0))
+
+
+# Map of common device classes to specific sensor types
+SENSOR_TYPE_MAP: dict[
+    SensorDeviceClass | str | None, type[CombinedEnergyReadingsSensor]
+] = {
+    SensorDeviceClass.ENERGY: EnergySensor,
+    SensorDeviceClass.POWER: PowerSensor,
+    SensorDeviceClass.WATER: WaterVolumeSensor,
+    SensorDeviceClass.POWER_FACTOR: PowerFactorSensor,
+}
```

Comment:
```suggestion
```

Instead of the map please use something similar to https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifi/sensor.py#L118